src/login/connect_server.h

```diff
@@ -0,0 +1,2257 @@
+    }
+
+    ShowDebug(fmt::format("char<{}> successfully saved", createchar.m_name));
+    return 0;
+};
+
+// Interaction with xiloader, port 54231
+class auth_session : public handler_session
+{
+public:
+    auth_session(asio::ssl::stream<asio::ip::tcp::socket> socket)
+    : handler_session(std::move(socket))
+    {
+        DebugSockets("auth_session");
+    }
+
+public:
+    void start()
+    {
+        auto self(shared_from_this());
+        // clang-format off
+
+        socket_.async_handshake(asio::ssl::stream_base::server,
+        [this, self](std::error_code ec)
+        {
+            if (!ec)
+            {
+                do_read();
+            }
+            else
+            {
+                ShowWarning(fmt::format("Error: ({}), {}", ec.value(), ec.message()));
+                ShowWarning("Failed to handshake!");
+                socket_.next_layer().close();
+            }
+        });
+        // clang-format on
+    }
+
+protected:
+    void do_read()
+    {
+        auto self(shared_from_this());
+        // clang-format off
+        socket_.async_read_some(asio::buffer(data_, max_length),
+        [this, self](std::error_code ec, std::size_t length)
+        {
+            if (!ec)
+            {
+                read_func();
+            }
+            else
+            {
+                DebugSockets(ec.message());
+                handle_error(ec, self);
+            }
+        });
+        // clang-format on
+    }
+
+    void read_func() override
+    {
+        auto newModeFlag = ref<uint8>(data_, 0) == 0xFF;
+        if (!newModeFlag)
+        {
+            ShowDebug("Old xiloader connected. Not supported.");
+            ref<uint8>(data_, 0) = LOGIN_ERROR;
+            do_write(1);
+            return;
+        }
+
+        /*auto componentFlags = ref<uint64>(data_, 1);
+
+        if (componentFlags & )
+        {
+            // ...
+        }*/
+
+        char usernameBuffer[17] = {};
+        char passwordBuffer[17] = {};
+
+        std::memcpy(usernameBuffer, data_ + 0x09, 16);
+        std::memcpy(passwordBuffer, data_ + 0x19, 16);
+
+        // std::string reservedData(data_ + 0x40, 17);
+        std::string version(data_ + 0x51, 5);
+
+        std::string username(usernameBuffer, 16);
+        std::string password(passwordBuffer, 16);
+
+        if (strncmp(version.c_str(), SUPPORTED_XILOADER_VERSION, 5) != 0)
+        {
+            ref<uint8>(data_, 0) = LOGIN_ERROR_VERSION_UNSUPPORTED;
+
+            do_write(1);
+            return;
+        }
+
+        int8 code = ref<uint8>(data_, 0x29);
+
+        DebugSockets(fmt::format("auth code: {}", code));
+
+        // data check
+        if (check_string(username, 16) && check_string(password, 16))
+        {
+            ShowWarning("login_parse: send unreadable data");
+            return;
+        }
+        auto sql = std::make_unique<SqlConnection>();
+
+        char escaped_name[16 * 2 + 1] = {};
+        char escaped_pass[32 * 2 + 1] = {};
```

Comment:
Just as I say that, theres a 16/32 discrepancy here?

Reply:
password is 32 characters now